posix: pthread: correct race condition between pthread_create and pthread_join

cfriedt opened the pull request:
Recently, a race condition was discovered in `pthread_create()` and `pthread_join()` that would trigger an assertion in `kernel/sched.c` of the form below.
```
aborted _current back from dead
```

This required a few changes to address thoroughly, but all indications point to a significant improvement in determinism and stability.

* revert 77e3e8961d1162a839665aaa244c5c47189b896d
* ensure final pthread state is `PTHREAD_TERMINATED`
* add `z_thread_wake_joiners()` (c/o @andyross)
* fix create/join race in thread lifecycle (c/o @andyross)
* add pthread create / join stress test

Fixes #56163

Andy's contributions brought in from #57721 (I hope that's ok). Thanks boss 👍 

**Compliance issue is a false positive**

Commit:
tests: posix: stress test for pthread_create and pthread_join

Recently, a race condition was discovered in `pthread_create()`
and `pthread_join()` that would trigger an assertion in
`kernel/sched.c` of the form below.
```
aborted _current back from dead
```

Add a test that stresses the pthread implementation
in order to more easily reproduce the failure.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
Revert "posix: pthread: consider PTHREAD_EXITED state in pthread_create"

This reverts commit da0398d19879a7548ea6cf36e0145930c50a0ce7.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: initial and final state must be TERMINATED

Previously, we considered `PTHREAD_EXITED` as a valid state
for reclaiming a thread. That was not the case.
`PTHREAD_EXITED` requires that `pthread_join()` is called while
the thread is in `PTHREAD_EXITED` state.

With that the calling thread may retrieve the return value of
the joining thread.

Additionally, correct incorrectly aliased pointer argument to
`k_thread_abort()`.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

evgeniy-paltsev commented on tests/posix/pthread_pressure/src/main.c:
```diff
@@ -0,0 +1,94 @@
+LOG_MODULE_REGISTER(pthread_pressure, LOG_LEVEL_INF);
+
+/* bitmask of available threads */
+static atomic_t free_threads =
+	ATOMIC_INIT((NUM_THREADS == 32) ? UINT32_MAX : BIT_MASK(NUM_THREADS));
```
As I can see we can simplify it to:

```suggestion
static atomic_t free_threads = ATOMIC_INIT(BIT_MASK(NUM_THREADS);
```

cfriedt replied:
Fixed! Thank you

andyross committed:
lib/pthread: Fix create/join race in thread lifecycle

The pthread_join() implementation was done using a condition variable
and not k_thread_join(), and that's actually fatal.  It's not possible
in Zephyr to guarantee a thread has exited in a race-free way without
calling either k_thread_abort() or k_thread_join().

This opened a hole where an app could call pthread_join(), which would
go to sleep waiting on the condition variable, which would be signaled
from the exit path of the thread, and then return to call
pthread_create() (and thus k_thread_create()) on the same pthread_t
object (and thus the same struct k_thread) BEFORE THE EARLIER THREAD
HAD ACTUALLY EXITED.

This makes the scheduler blow up, at least on SMP (I wasn't personally
able to make it happen on uniprocessor configs, though the race is
present always).

Rework to call k_thread_join().  There's no other correct way to do
this.

Fixes #56163

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
kernel: Add z_thread_wake_joiners() utility

POSIX has a somewhat odd behavior that when you call
pthread_detach()[1], it is supposed to synchronously wake up anyone
currently joining the thread.  That isn't a Zephyr kernel behavior[2],
so it needs to be exposed.

The implementation is as simple as it sounds.  This is minimal: we
don't try to keep any state around to inform the k_thread_join() call
whether the thread actually exited or if this new function was used.
That has to be provided by the outer layers (pthreads already do
this).

[1] Which tells the OS/platform that it can clean the thread up as
soon as it exits and doesn't have to keep it around as a zombie until
someone calls pthread_join() on it.

[2] Because... why!?  Not needing to keep a thread around for joiners
isn't the same thing as disallowing joins!

Signed-off-by: Andy Ross <andyross@google.com>

cfriedt commented on lib/posix/posix_internal.h:
```diff
@@ -31,10 +31,10 @@ enum pthread_state {
 	/* The thread is running and joinable. */
 	PTHREAD_JOINABLE = PTHREAD_CREATE_JOINABLE,
+	/* A joinable thread exited and its return code is available. */
+	PTHREAD_EXITED,
```
This is mainly for the mental model (if TERMINATED comes below EXITED, it's a bit more intuitive)

jgl-meta reviewed (dismissed):
I don't see anything obviously incorrect.

keith-packard commented on lib/posix/posix_internal.h:
```diff
@@ -31,10 +31,10 @@ enum pthread_state {
 	PTHREAD_JOINABLE = PTHREAD_CREATE_JOINABLE,
+	/* A joinable thread exited and its return code is available. */
+	PTHREAD_EXITED,
 	/* The thread structure is unallocated and available for reuse. */
```
This name doesn't match the described semantic very well; is this more like `PTHREAD_NOT_IN_USE` or `PTHREAD_IDLE`?

cfriedt replied:
Agreed. There is a distinct ambiguity between `TERMINATED` and `EXITED`. For the time being, I'd prefer to keep naming though, to ensure that backports are consistent.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -177,7 +176,7 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
 	for (pthread_num = 0;
 	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
 		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
```
yup, this seems like a good change. It does mean that you cannot re-use a thread until it has been joined again, right?

cfriedt replied:
Correct

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -191,7 +190,7 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
 	if (rv != 0) {
 		key = k_spin_lock(&pthread_pool_lock);
-		thread->state = PTHREAD_EXITED;
+		thread->state = PTHREAD_TERMINATED;
```
Yes, this looks like there was just a bug here -- there's no way you'd be able to join this thread as it was never returned, right?

cfriedt replied:
Previously I added a change that allowed EXITED as well. But yeah, there was no previous documentation. I would definitely like to add some, but that is not as high priority

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -274,7 +272,6 @@ int pthread_cancel(pthread_t pthread)
 		} else {
 			thread->retval = PTHREAD_CANCELED;
 			thread->state = PTHREAD_EXITED;
-			pthread_cond_broadcast(&thread->state_cond);
```
Why was this condition variable even present? Was there some state transition that couldn't be monitored from the OS thread object itself?

cfriedt replied:
The original implementation of pthreads were almost completely orthogonal to k_thread.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -411,11 +407,12 @@ void pthread_exit(void *retval)
-	pthread_mutex_destroy(&self->state_lock);
 
-	pthread_cond_destroy(&self->state_cond);
+	if (self->state == PTHREAD_TERMINATED) {
```
yup, retain the ability to lock access to the thread itself.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -411,11 +407,12 @@ void pthread_exit(void *retval)
+	}
 
-	k_thread_abort((k_tid_t)self);
+	k_thread_abort(&self->thread);
```
nice type cleanup.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
-	pthread_mutex_lock(&pthread->state_lock);
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
```
this seems like an odd choice -- did the old code really not use `k_thread_join` at all?

cfriedt replied:
Yeah... again pthreads were quite different. Almost like "Linux threads" if you remember the old user mode thread implementation that came before nplt

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (pthread->state == PTHREAD_DETACHED) {
```
This is fetched outside of the lock; I think this only assumes state can be read atomically, right? Threads can't move from DETATCHED to some state other than TERMINATED?

cfriedt replied:
Yeah, i noticed that too. @andyross - any thoughts here?

andyross replied:
State is a word, so in practice it's atomic.  But yeah, outside the lock that gets squishy and we should tighten it up, which is a little complicated given that these states are just recapitulating the (separately synchronized!) kernel internal state.  Maybe a separate boolean/flag for "detached" would be easier to manage, since it can change in only one direction.

Not sure how that race can get you back to the scheduler crash, though.  That should only happen if you have a k_thread_create() call that begins before some context returns from k_thread_abort() or k_thread_join().  Maybe the test has a state where it ignores the error from pthread_join() (where it fails to join a detached thread) and is trying to recreate it before it exits?  I don't think pthreads actually gives you an API to handle that safely at all, since you can't use join...

cfriedt replied:
Used a local bool.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
 	pthread_mutex_unlock(&pthread->state_lock);
-	if (pthread->state == PTHREAD_EXITED) {
+
+	if (ret == 0 && pthread->state == PTHREAD_TERMINATED) {
```
I think this check is protecting the mutex destroy from happening twice because the thread was already TERMINATED, right? Might add a comment to make it clear, or even use a separate local boolean set in the code above which sets the state.

cfriedt replied:
Yes. A detached thread must destroy its own resources in exit, while a joinable thread must destroy resources in join.

cfriedt replied:
Added a comment.

keith-packard reviewed (commented):
This is looking like a nice cleanup -- well reasoned state transitions for threads that don't skip steps. I could have used a bit of documentation about the possible set of state transitions where those are defined. I'm also unsure about the use of the mutex in one case that appears like it could happen when the state was TERMINATED.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
 
+	k_thread_join(&pthread->thread, K_FOREVER);
+
+	pthread_mutex_lock(&pthread->state_lock);
```
Actually, the mutex could still be valid even if the state is terminated. It's ... not perfect yet.

cfriedt commented:
* added comments around `pthread_mutex_destroy()` in `pthread_join()` and `pthread_exit()`

cfriedt commented:
* used local `bool destroyed` in `pthread_join()` and `pthread_exit()`

cfriedt commented:
* call `pthread_join()` from `pthread_detach()` if thread has already exited (to ensure resources are destroyed)

keith-packard commented:
I like the `destroy` booleans, and just using `pthread_join` to clean up an exited thread from `pthread_detach`.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -436,24 +437,28 @@ int pthread_join(pthread_t thread, void **status)
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (pthread->state == PTHREAD_DETACHED) {
+		return EINVAL;
 	}
 
+	k_thread_join(&pthread->thread, K_FOREVER);
+
+	pthread_mutex_lock(&pthread->state_lock);
```
This part is still kind of bothering me..

keith-packard replied:
Could you use atomics for the state instead of a mutex?

keith-packard replied:
Alternatively, don't free the mutex ever. Use a static initializer for the thread pool. Or use a single mutex for all threads.

cfriedt replied:
Should be fixed with the latest rev. And actually pretty tight, I hope.

Commit:
posix: pthread: rewrite of create, join, detach, cancel, and exit

Previously, the pthread implementation used a `pthread_mutex_t`
to synchronize thread state. This was required because a
`pthread_cond_t` was used to notify other pthreads that were
attempting to join or exit the same pthread.

With the previous removal of the `pthread_cond_t` from
`struct posix_thread`, and the addition of the
`z_thread_wake_joiners()` utility, there is no longer a need
to utilize a `pthread_mutex_t` to synchronize state.

Instead, use a `struct k_spinlock` for synchronization and
`z_thread_wake_joiners()` for notification.

Additionally, use a `bitarray` instead of the the
`pthread_pool_lock`, which makes `pthread.c` resource
allocation more like that of the other pthread objects.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
Revert "tests: posix: pthread: init pthread_attr_t on each iteration"

This reverts commit 7c17bda3c23ed1ea33aa65df2732ac9a586640a4.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: posix: common: add a small delay in pthread_descriptor_leak

Technically speaking, there should be a synchronization to ensure
that a spawned thread has started before `pthread_join()` is
called.

However, since we're recycling an existing no-op thread, just add
a small delay.

A separate test will be used to demonstrate stressing
`pthread_create()` and `pthread_join()` which will use a
synchronization primitive.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* use a `struct k_spinlock` instead of a `pthread_mutex_t` to synchronize pthread internal state
* remove `pthread_pool_lock`
* use `atomic_cas()` to select next available pthread

A positive side-effect is that now each `pthread_t` does not consume a `pthread_mutex_t` (or a `pthread_cond_t` since @andyross's change). Both `pthread_mutex_t` and `pthread_cond_t` are limited in number by `Kconfig`.

cfriedt commented:
@andyross - I hope you'll find pthreads a bit more pleasant to look at with this last rev

andyross reviewed (dismissed):
This all makes sense to me

keith-packard commented on lib/posix/posix_internal.h:
```diff
@@ -31,10 +32,10 @@ enum pthread_state {
+	/* A joinable thread exited and its return code is available. */
+	PTHREAD_EXITED,
 	/* The thread structure is unallocated and available for reuse. */
 	PTHREAD_TERMINATED,
```
If this constant was set equal to zero, would that allow the structure to not require any initialization before first use?

cfriedt replied:
> If this constant was set equal to zero, would that allow the structure to not require any initialization before first use?

I opted to have the CREATE constants align with the states across all libcs

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -143,6 +142,31 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+static bool pthread_attr_is_valid(const struct pthread_attr *attr)
+{
+	/*
+	 * FIXME: Pthread attribute must be non-null and it provides stack
```
Hrm. This seems a bit unfortunate. Could the stack be dynamically allocated from the user heap if a user heap were available?

evgeniy-paltsev replied:
@keith-packard please note that we already has this restriction (rather than introducing it in this PR) - so it shouldn't be fixed by this PR.

cfriedt replied:
> Hrm. This seems a bit unfortunate. Could the stack be dynamically allocated from the user heap if a user heap were available?

That is the topic of a longstanding issue & series of PRs I've been working on (in my "spare time") for... "a while" 😅

Truthfully, this should be handled at the Zephyr level and not at the pthreads level.

Currently blocked on MMU issues.

keith-packard commented on lib/posix/posix_internal.h:
```diff
@@ -51,10 +52,9 @@ struct posix_thread {
-	pthread_mutex_t state_lock;
-	pthread_cond_t state_cond;
+	/* Pthread State (of type enum pthread_state) */
+	atomic_t state;
```
Given that "most" accesses to this occur under the lock, would it be more efficient to require all of them to be under the lock and avoid using an atomic here? Atomic operations are quite expensive on some architectures, and not available on others (cortex M0, I'm looking at you).

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -143,6 +142,31 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	}
+
+	/* require a valid detachstate */
+	if (!(attr->detachstate == PTHREAD_JOINABLE || attr->detachstate == PTHREAD_DETACHED)) {
```
Do we need this here? Or is it sufficient to validate the parameter to `pthread_attr_setdetachstate`?

cfriedt replied:
I could probably change it to `__ASSERT_NO_MSG()`

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -154,70 +178,42 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
+		if (atomic_cas(&thread->state, PTHREAD_TERMINATED, attr->detachstate)) {
+			sys_slist_init(&thread->key_list);
```
I think this call to `sys_slist_init` can be placed outside the loop with the rest of the thread initialization bits.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -268,15 +265,14 @@ int pthread_cancel(pthread_t pthread)
-		pthread_mutex_lock(&thread->state_lock);
+		state_key = k_spin_lock(&thread->state_lock);
 		if (thread->state == PTHREAD_DETACHED) {
 			thread->state = PTHREAD_TERMINATED;
```
yeah, this looks like a locking issue -- you're setting the state to TERMINATED while holding the spinlock, but the above allocation code isn't holding the lock when transitioning from TERMINATED to a running state. I think you need to either do all state manipulations using atomics or all state manipulations under the spin lock. And, I think the latter option will be easier to reason about.

cfriedt replied:
Hmm.. yea, I could see switching to only spinlocks.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -268,15 +265,14 @@ int pthread_cancel(pthread_t pthread)
 		} else {
 			thread->retval = PTHREAD_CANCELED;
 			thread->state = PTHREAD_EXITED;
-			pthread_cond_broadcast(&thread->state_cond);
```
What was this broadcast used for? And, if necessary, how is it being replaced?

cfriedt replied:
This was removed by Andy. I assumed that `k_thread_join()` would notify other waiting threads. I'll double check.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -436,26 +428,37 @@ int pthread_join(pthread_t thread, void **status)
+		ret = k_thread_join(&pthread->thread, K_FOREVER);
+		__ASSERT(ret == 0, "k_thread_join() failed: %d", ret);
+		state_key = k_spin_lock(&pthread->state_lock);
+		if (ret != 0) {
```
You might need to go back and re-validate the current `pthread->state` value to make sure the value is as expected; trying to reason about all possible transitions that can happen between the unlock and lock is hard.

cfriedt replied:
Nothing else would set the state here. It would actually be the responsibility of this function to set the state to `PTHREAD_EXITED`. However, since the state would be changed again only a few lines below, changing it here would be redundant.

I could add that line of code for clarity. Perhaps the compiler would optimize it out in some cases.

keith-packard replied:
You're unlocking the state_lock, so couldn't another thread call a thread API and mangle things before you manage to re-lock it?

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -718,7 +720,7 @@ static int posix_thread_pool_init(void)
 
 	for (i = 0; i < CONFIG_MAX_PTHREAD_COUNT; ++i) {
-		posix_thread_pool[i].state = PTHREAD_EXITED;
+		posix_thread_pool[i].state = PTHREAD_TERMINATED;
```
as above, if `PTHREAD_TERMINATED` were to be assigned the value 0, then this wouldn't be necessary.

cfriedt replied:
> as above, if `PTHREAD_TERMINATED` were to be assigned the value 0, then this wouldn't be necessary.
> 

IIRC it was changed so that there was agreement between detached state and joinable among the various libcs

keith-packard replied:
Ah, that's slightly annoying, but understandable. Probably better to have this initialized here than stick the thread pool in data with a static initializer.

keith-packard commented on tests/posix/pthread_pressure/src/main.c:
```diff
@@ -0,0 +1,98 @@
+LOG_MODULE_REGISTER(pthread_pressure, LOG_LEVEL_INF);
+
+/* bitmask of available threads */
+static atomic_t free_threads = ATOMIC_INIT(BIT_MASK(NUM_THREADS));
```
Could use a set of in-use threads to allow 0 initializer here.

cfriedt replied:
I have a "setup" function now too. Probably not worth it to split this hair.

keith-packard commented on tests/posix/pthread_pressure/src/main.c:
```diff
@@ -0,0 +1,98 @@
+		maybe_free_threads = (uint32_t)atomic_get(&free_threads);
+		LOG_DBG("free_threads: %x", maybe_free_threads);
+		while (maybe_free_threads != 0) {
+			i = find_lsb_set(maybe_free_threads) - 1;
```
This looks like it will re-use the LSB value and possibly never hit other threads? Maybe `pthread_fun` should use `k_busy_wait` to spin the CPU for a while (random?) so that threads are actually active for some time?

cfriedt replied:
> This looks like it will re-use the LSB value and possibly never hit other threads? Maybe `pthread_fun` should use `k_busy_wait` to spin the CPU for a while (random?) so that threads are actually active for some time?

The point of the test was to not to introduce any unnecessary delays between creation and join.

The only necessary delay was to ensure that the thread is spawned before joining as soon as possible.

Technically speaking, this could also be solved by introducing another state.

keith-packard replied:
I don't think we need any more state in the pthread implementation; we just need to ensure that serialized calls between `k_thread_create` and `k_thread_join` are always valid, no matter what the created thread is doing.

cfriedt replied:
As it turns out, even those calls outside of pthreads are racey 😞

keith-packard reviewed (commented):
I think switching from atomic_t for the state and evaluating under the per-thread lock will end up with fewer potential issues. Otherwise, just a few minor questions. This is looking a *lot* cleaner.

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -587,12 +587,15 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
-		zassert_ok(pthread_attr_setstack(&attr, &stack_e[0][0], STACKS));
 		zassert_ok(pthread_create(&pthread1, &attr, create_thread1, NULL),
 			   "unable to create thread %zu", i);
+		/* Small delay to prevent joining before the thread has been spawned */
```
> This looks like a possible source of application bugs if it is actually required here. I would remove this sleep and make sure the implementation was actually capable of doing back-to-back create/join calls.

IIRC, even on Linux there is no guarantee that a thread is spawned before `pthread_create()` returns.

My initial respin of this rev did introduce a semaphore to block the calling thread until the spawned thread had started. However, that would also punish the threads that do not require that bit of synchronization so I removed it.

keith-packard replied:
yeah, you definitely don't want to serialize thread startup. However, given that `pthread_create` doesn't return until after `k_thread_create` has been called, I would think it should be valid to call `k_thread_join` any time after `k_thread_create` has returned. If not, then I think we've got some sharp edges in the API that are going to catch developers by surprise.

andyross replied:
That's indeed the behavior: you can call k_thread_join() at any point after k_thread_create() has returned in any thread, even after it has aborted.   And FWIW: even outside that regime you're safe as long as the k_thread struct is zero-initialized (joining garbage data is, heh, undefined).

cfriedt replied:
> yeah, you definitely don't want to serialize thread startup. However, given that `pthread_create` doesn't return until after `k_thread_create` has been called, I would think it should be valid to call `k_thread_join` any time after `k_thread_create` has returned. If not, then I think we've got some sharp edges in the API that are going to catch developers by surprise.

If that should be the expected behaviour, even though it differs from other POSIX systems, then we would likely need to hold a spinlock until after `k_thread_create()` returns.

That might be fine

keith-packard replied:
Maybe I'm just misunderstanding something here. I believe, reading the pthreads POSIX spec, that this is valid and well defined code:
```
    pthread_t th;
    pthread_create(&th, NULL, thread_start, NULL);
    pthread_join(th, NULL);
```
Or am I just confused about what we're discussing?

As for holding the spinlock through `k_thread_create()`, I suspect that's a very bad idea as spinlocks will likely block interrupts on the current core, right?

keith-packard replied:
From my reading of your current version, I think you'd only need additional synchronization if the underlying `k_thread_create()` and `k_thread_join()` functions aren't safe to invoke back-to-back from the parent thread. You've done a good job of making the pthread implementation a pretty thin wrapper on top of the underlying Zephyr thread implementation, just a bit of state to deal with the extra POSIX semantics.

cfriedt replied:
I think it's best to aim for the ideal - so users should feel comfortable that they can busy loop on calling `k_thread_create()` and `k_thread_join()` in quick succession, even if it's not the smartest thing to do.

Unfortunately, the assumptions are not sound. This may require a bit of additional effort on the kernel side.

evgeniy-paltsev commented on tests/posix/pthread_pressure/Kconfig:
```diff
@@ -0,0 +1,38 @@
+config TEST_DURATION_S
+	int "Number of seconds to run the test"
+	range 1 21600
+	default 1
```
I'd bamp the default as we want to have this 'torture' test to be executed at least for some time. Something like 60 seconds looks more suitable here.

cfriedt replied:
60s would even be quite a lot for CI runs.

Any thoughts on that @stephanosio

stephanosio replied:
60s is fine. We have plenty of tests that take just as long or longer.

cfriedt commented:
@keith-packard, @andyross, and also @jgl-meta - can you please sanity check this for me?

```mermaid
---
title: Pthread States
---
stateDiagram-v2
    [*] --> TERMINATED
    TERMINATED --> DETACHED: create()
    DETACHED --> ABORTING: exit() 
```

cfriedt commented:
@keith-packard, @andyross, and also @jgl-meta - can you please sanity check this for me?

TL;DR - I think the state lock must be a semaphore and cannot be a mutex or spinlock if we want to leverage `k_thread_join()` and `k_thread_abort()` (which we do).

```mermaid
---
title: Proposed Pthread States
---
stateDiagram-v2
    [*] --> TERMINATED
    %% Initially detached threads
    TERMINATED --> DETACHED: create()
    DETACHED --> ABORTING: exit()
    DETACHED --> ABORTING: cancel()
    ABORTING --> TERMINATED: work()
    %% Initially joinable threads
    TERMINATED --> JOINABLE: create()
    JOINABLE --> DETACHED: detach()
    JOINABLE --> ABORTING: exit()
    JOINABLE --> ABORTING: cancel()
    ABORTING --> EXITED: work()
    EXITED --> TERMINATED: join()
```

**Long version**

In order to fully eliminate races, we need to add another state (`ABORTING` below) that actually does the `k_thread_abort()` with the lock held in a separate work queue thread since letting go of the lock would necessarily introduce races. Lastly, I think there only needs to be a single lock in `struct posix_thread` that synchronizes both state and cancel state. More than one lock seems like overkill.

cfriedt commented:
> However, when looking at it, I'm confused as to why you think you need a new state here -- `work()` takes ABORTING to either `EXITED` or `TERMINATED`, and `EXITED` only goes to `TERMINATED`, so it seems like you could merge ABORTING and TERMINATED.

Yeah, I think a flag would be appropriate (e.g. `abort_pending`), because otherwise we would need to remember the previous state, which is not as fun. We can represent a separate "state" when that flag is set.

> Maybe the actions that cause the state transitions need some additional annotation so we can understand the difference between the arcs with the same operation coming from the same node? For instance, TERMINATED to DETACHED/JOINABLE might be labeled as create(DETACHED)/create(JOINABLE)? I'd like to know how the two transitions from ABORTING are different.

The create parameter was there originally, but I just thought it was obvious. Will update. Same with transitions from ABORTING - a thread only goes to EXITED if the previous state was JOINABLE.

> I'm hesitant to encourage a move from a spinlock/mutex to a semaphore as semaphores are harder to reason about.

Hopefully I can keep it simple.

> Holding the lock across thread transitions is kinda icky. The "usual" way to manage things like this is to simply re-validate the state in the second thread and work from there, potentially re-trying the operation, potentially performing some other operation

I'm not convinced that is sufficient to avoid race conditions and so far, my experiments show that hypothesis to be true.

keith-packard commented:
> I'm not convinced that is sufficient to avoid race conditions and so far, my experiments show that hypothesis to be true.

Yeah, I suspect we'll have to see the implementation to get to ground truth here. Diagrams are all nice and everything, but they don't actually run on computers...

cfriedt commented:
@keith-packard , @andyross - some good / bad news. I improved my testsuite to also test raciness between `k_thread_create()` and `k_thread_join()` as well.

The good news, is that pthreads are not alone in that they are racey 😀 

The bad news, is that the same race exists today with k_threads 😞 

Currently, testing pthreads is disabled by default. I also reduced the test duration from 60s to 10s. 

Any failures in the `pthread_pressure` suite are exclusively with k_threads.

I was really hoping to get this fixed for v3.4.0 (and also v2.7.5) - but I'm a bit exhausted at the moment.

Sadly, the "aborted _current back from dead" issue exists with all libcs in various configurations and on multiple architectures.

```shell
twister -T tests/posix/pthread_pressure/
Renaming output directory to /home/cfriedt/zephyrproject/zephyr/twister-out.3
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.3.0-4131-g1767a0844e0a
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
ERROR   - qemu_x86_64               tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_x86_64/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log
ERROR   - qemu_x86_64               tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_x86_64/tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc/handler.log
ERROR   - qemu_cortex_a53_smp       tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc/handler.log
ERROR   - qemu_cortex_a53_smp       tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log
ERROR   - qemu_riscv64_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log
ERROR   - qemu_riscv32_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv32_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log
ERROR   - qemu_cortex_a53_smp       tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls/handler.log
ERROR   - qemu_cortex_a53_smp       tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib/handler.log
ERROR   - qemu_x86_64               tests/posix/pthread_pressure/portability.posix.pthread_pressure  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_x86_64/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log
ERROR   - qemu_cortex_a53_smp       tests/posix/pthread_pressure/portability.posix.pthread_pressure  FAILED : unexpected eof
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log
ERROR   - qemu_riscv64_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc/handler.log
ERROR   - qemu_riscv32_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv32_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls/handler.log
ERROR   - qemu_riscv64_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib/handler.log
ERROR   - qemu_riscv32_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv32_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log
ERROR   - qemu_riscv64_smp          tests/posix/pthread_pressure/portability.posix.pthread_pressure  FAILED : Timeout
ERROR   - see: /home/cfriedt/zephyrproject/zephyr/twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log
INFO    - Total complete:  275/ 275  100%  skipped:  171, failed:   15, error:    0
INFO    - 9 test scenarios (275 test instances) selected, 171 configurations skipped (163 by static filter, 8 at runtime).
INFO    - 89 of 275 test configurations passed (85.58%), 15 failed, 0 errored, 171 skipped with 0 warnings in 157.44 seconds
INFO    - In total 121 test cases were executed, 429 skipped on 40 out of total 571 platforms (7.01%)
INFO    - 102 test configurations executed on platforms, 2 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests/posix/pthread_pressure/portability.posix.pthread_pressure on qemu_cortex_a53_smp failed (unexpected eof)
INFO    - 2) tests/posix/pthread_pressure/portability.posix.pthread_pressure on qemu_riscv64_smp failed (Timeout)
INFO    - 3) tests/posix/pthread_pressure/portability.posix.pthread_pressure on qemu_riscv32_smp failed (Timeout)
INFO    - 4) tests/posix/pthread_pressure/portability.posix.pthread_pressure on qemu_x86_64 failed (unexpected eof)
INFO    - 5) tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib on qemu_cortex_a53_smp failed (unexpected eof)
INFO    - 6) tests/posix/pthread_pressure/portability.posix.pthread_pressure.newlib on qemu_riscv64_smp failed (Timeout)
INFO    - 7) tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls on qemu_cortex_a53_smp failed (unexpected eof)
INFO    - 8) tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls on qemu_riscv32_smp failed (Timeout)
INFO    - 9) tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib on qemu_cortex_a53_smp failed (unexpected eof)
INFO    - 10) tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib on qemu_riscv64_smp failed (Timeout)
INFO    - 
INFO    - To rerun the tests, call twister using the following commandline:
INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
INFO    - 
INFO    - west twister -p qemu_riscv64_smp -s tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib
INFO    - or with west:
INFO    - west build -p -b qemu_riscv64_smp -T tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - Run completed
```

```shell
$ grep "back from dead" $(find twister-out/ -name handler.log)
twister-out/qemu_riscv32_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log:    aborted _current back from dead
twister-out/qemu_riscv32_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log:       aborted _current back from dead
twister-out/qemu_x86_64/tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc/handler.log:   aborted _current back from dead
twister-out/qemu_cortex_a53_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure/handler.log:    aborted _current back from dead
twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.tls.newlib/handler.log:    aborted _current back from dead
twister-out/qemu_riscv64_smp/tests/posix/pthread_pressure/portability.posix.pthread_pressure.picolibc/handler.log:      aborted _current back from dead
```

keith-packard commented:
A serious bummer. Sounds like we need to address the `k_thread` races before you can even begin to fix the pthreads issues (if any remain).

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -154,70 +178,54 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
 
-	pthread_cond_init(&thread->state_cond, &cond_attr);
-	sys_slist_init(&thread->key_list);
+		if (found) {
```
this looks redundant with the loop condition?

cfriedt replied:
Fixed - thanks for spotting that.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -154,70 +178,54 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	}
 
 	*newthread = pthread_num;
 	k_thread_create(&thread->thread, attr->stack, attr->stacksize,
```
I think this is OK -- you're leaving a gap in the locking between the loop exit and this point, but the thread value itself isn't visible to any external users, so it can't be passed back to another thread API.

However, you might want to move the assignment to newthread until after k_thread_create returns in case that's a global location visible to another thread.

cfriedt replied:
Probably smart.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -254,34 +262,43 @@ int pthread_setcancelstate(int state, int *oldstate)
-	k_spinlock_key_t cancel_key;
 
-	if ((thread == NULL) || (thread->state == PTHREAD_TERMINATED)) {
+	if (thread == NULL) {
```
yeah, there's no protection here for using a thread index which is stale; checking only for one possible stale state (TERMINATED) provides a false sense of security. Makes me wonder if there should be some kind of validating mode for the API where a generation number is encoded in the upper bits of the thread ID that are validated against a current generation number stored in the thread?

cfriedt replied:
I like the thought and it would at least replicate some of the other validation schemes in e.g. `pthread_mutex.c`.

To be honest, I think I should be able to transition pthreads to use `<zephyr/sys/bitarray.h>` as well. 

Let me give it another spin.

Also, thanks @andyross and @npitre for validating the issue on aarch64. Will likely use an `arch_exclude` there if I am able to.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -377,28 +394,36 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
-	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
-		self->cancel_state = PTHREAD_CANCEL_ENABLE;
-	}
+	__ASSERT(self != NULL, "not a valid pthread (%p)", k_current_get());
```
Hrm. Is pthread_t at least as large as size_t and unisigned? Otherwise, I think `to_posix_thread()` needs some better bounds checking to avoid possible wrap issues (imagine a 32-bit pthread_t on a  64-bit size_t system, it's possible that a large address difference will wrap around back into range).

cfriedt replied:
`pthread_t` is a `uint32_t` in Zephyr, so should be safe.

keith-packard replied:
that makes it unsafe -- on 64 bit targets, size_t is 64 bits, so you will get aliasing when converting from size_t to pthread_t. Probably need to fix pthread_self to bounds check the computed pointer before converting to an array index with the subtraction.

cfriedt replied:
True - for some reason I was thinking it was 32-bit. I'll add a bounds check in this rev

andyross commented:
Repeating the comment from elsewhere, regarding the kernel bug tracked in #58116 : FWIW: it seems not to be a kernel race in thread lifecycle, more like a arm64-specific race in context handling (and maybe x86_64, but with 1000x lower frequency).

FWIW: I'd treat this as two separate issues and merge the pthread fixes. If you need to tune the test[1] on arm64 for CI, that seems acceptable as long as we have a bug tracking the instability.

[1] SMP tests are already retried in CI because of unavoidable timer skew problems with scheduling "CPUs" onto host threads that can be preempted for 10ms+ in busy CI runs. (Which isn't what we're seeing here -- no timer interaction in this test!). The frequency I'm seeing probably wouldn't even show up above the noise floor.

cfriedt commented:
I'm going to close this PR, as somehow pthreads is less reliable than before, in terms of passing Twister builds, but I'll make a separate PR for the `pthread_pressure` test and `pthread_descriptor_leak` tests.
